pkg/observer/observer_stats.go

```diff
@@ -41,44 +41,49 @@ func (s *statValue) GetValuePtr() unsafe.Pointer {
 }
 func (s *statValue) DeepCopyMapValue() bpf.MapValue {
 	v := &statValue{
-		Value: make([]int64, runtime.NumCPU()),
+		Value: make([]int64, len(s.Value)),
 	}
 	copy(v.Value, s.Value)
 	return v
 }
 
-func (k *Observer) startUpdateMapMetrics() {
-	update := func() {
-		for _, m := range sensors.AllMaps {
-			pin := filepath.Join(option.Config.MapDir, m.Name)
-			pinStats := pin + "_stats"
+func updateMapMetric(name string) {
+	pin := filepath.Join(option.Config.MapDir, name)
+	pinStats := pin + "_stats"
 
-			mapLinkStats, err := bpf.OpenMap(pinStats)
-			if err != nil {
-				continue
-			}
-			mapLink, err := bpf.OpenMap(pin)
-			if err != nil {
-				continue
-			}
+	mapLinkStats, err := bpf.OpenMap(pinStats)
+	if err != nil {
+		return
+	}
+	defer mapLinkStats.Close()
+	mapLink, err := bpf.OpenMap(pin)
+	if err != nil {
+		return
+	}
+	defer mapLink.Close()
 
-			zeroKey := &statKey{}
-			value, err := mapLinkStats.Lookup(zeroKey)
-			if err != nil {
-				continue
-			}
+	zeroKey := &statKey{}
+	value, err := mapLinkStats.Lookup(zeroKey)
+	if err != nil {
+		return
+	}
 
-			v, ok := value.DeepCopyMapValue().(*statValue)
-			if !ok {
-				continue
-			}
-			sum := int64(0)
-			for cpu := int(0); cpu < runtime.NumCPU(); cpu++ {
-				sum += v.Value[cpu]
-			}
-			mapmetrics.MapSizeSet(m.Name, int(mapLink.MapInfo.MaxEntries), float64(sum))
-			mapLink.Close()
-			mapLinkStats.Close()
+	v, ok := value.DeepCopyMapValue().(*statValue)
```

Comment:
Do we need the deep copy here? It looks like we just sum the values. Could we cast instead and save a copy?

Reply:
I think you are right.
The deepcopy was in the original code though, and maybe this is worth doing as a followup.